lib/SILOptimizer/SemanticARC/CoroutineLifetimeExtender.cpp

```diff
@@ -0,0 +1,295 @@
+//===--- CoroutineLifetimeExtender.cpp ------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+///
+/// This pass contains a simple transform that extends the lifetime of a
+/// coroutine with a trivial resume to transitive uses of copies of yielded
+/// values from the coroutine and then eliminates those copies.
+///
+//===----------------------------------------------------------------------===//
+
+#include "Context.h"
+#include "OwnershipLiveRange.h"
+#include "Transforms.h"
+
+#include "swift/SIL/ApplySite.h"
+#include "swift/SIL/BasicBlockUtils.h"
+#include "swift/SIL/OwnershipUtils.h"
+#include "swift/SIL/SILBuilder.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SILOptimizer/PassManager/Passes.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/ValueLifetime.h"
+
+using namespace swift;
+using namespace swift::semanticarc;
+
+//===----------------------------------------------------------------------===//
+//                        Process Candidate Functions
+//===----------------------------------------------------------------------===//
+
+static bool processCandidates(Context &ctx,
+                              ArrayRef<FullApplySite> applySiteList) {
+  assert(!applySiteList.empty() && "Must have at least one candidate");
+
+  SmallVector<SILInstruction *, 32> scratchBuffer;
```

Comment:
This is not truly a "scratch" buffer because it's populated in one part of the code, used in other parts, and central to understanding the algorithm. I would call it something like lifetimeExtendingUses. This is a good place to comment that it will be used to compute the extended lifetime of the coroutine. Therefore it must contain the consuming uses of all the copies in addition to the current end_applies.